hw/drivers/uart/include/uart/uart.h

```diff
@@ -104,6 +98,14 @@ struct uart_conf {
     void *uc_cb_arg;
 };
 
+struct uart_dev {
+    struct os_dev ud_dev;
+    struct uart_driver_funcs ud_funcs;
+    struct uart_conf ud_conf;
```

Comment:
I would recommend remembering the fields from speed -> flow_ctl settings, instead of retaining the function pointers (which the hal driver will remember). This'll save some RAM. I think the call to hal_uart_init_cbs() from uart_hal_resume() is unnecessary.